src/irmin-containers/linked_log.ml

```diff
@@ -0,0 +1,230 @@
+(*
+ * Copyright (c) 2020 KC Sivaramakrishnan <kc@kcsrk.info>
+ * Copyright (c) 2020 Anirudh Sunder Raj <anirudh6626@gmail.com>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+open Lwt.Infix
+
+let return = Lwt.return
+
+type ('a, 'b, 'c) log_item = { time : 'a; msg : 'b; prev : 'c option }
+
+module Log_item (T : Time.S) (K : Irmin.Hash.S) (V : Irmin.Type.S) :
+  Irmin.Type.S with type t = (T.t, V.t, K.t) log_item = struct
+  type t = (T.t, V.t, K.t) log_item
+
+  let t =
+    let open Irmin.Type in
+    record "t" (fun time msg prev -> { time; msg; prev })
+    |+ field "time" T.t (fun r -> r.time)
+    |+ field "msg" V.t (fun r -> r.msg)
+    |+ field "prev" (option K.t) (fun r -> r.prev)
+    |> sealr
+end
+
+type ('a, 'b, 'c) store_item =
+  | Value of ('a, 'b, 'c) log_item
+  | Merge of ('a, 'b, 'c) log_item list
+
+module Store_item (T : Time.S) (K : Irmin.Hash.S) (V : Irmin.Type.S) :
+  Irmin.Type.S with type t = (T.t, V.t, K.t) store_item = struct
+  module L = Log_item (T) (K) (V)
+
+  type t = (T.t, V.t, K.t) store_item
+
+  let t =
+    let open Irmin.Type in
+    variant "t" (fun value merge ->
+      function Value v -> value v | Merge l -> merge l)
+    |~ case1 "Value" L.t (fun v -> Value v)
+    |~ case1 "Merge" (list L.t) (fun l -> Merge l)
+    |> sealv
+end
+
+module Log_ops
+    (C : Cas_maker.S)
+    (T : Time.S)
+    (K : Irmin.Hash.S)
+    (V : Irmin.Type.S) =
+struct
+  module Store = struct
+    include C.CAS_Maker (K) (Store_item (T) (K) (V))
+
+    let get_store =
+      let st = v @@ C.config in
+      fun () -> st
+  end
```

Comment:
I don't like staging mutable state inside modules like this, if at all possible. (I'm actually currently in the process of removing all occurrences of this pattern elsewhere in Irmin.) Is there any way we can avoid doing that here?

Reply:
Would this be okay? `let get_store () = v @@ C.config`

Reply:
I believe the "linked" property requires this to be the same store each time, so `(fun () -> v C.config)` would just be requiring `v` to keep this state internally instead (otherwise the stores aren't linked at all). See my [other comment](https://github.com/mirage/irmin/pull/1014#discussion_r450849752) for a different idea.